cns/requestcontroller/kubecontroller/crdreconciler.go

```diff
@@ -16,44 +19,132 @@ import (
 type CrdReconciler struct {
 	KubeClient KubeClient
 	NodeName   string
+	Namespace  string
 	CNSClient  cnsclient.APIClient
 }
 
 // Reconcile is called on CRD status changes
 func (r *CrdReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	var nodeNetConfig nnc.NodeNetworkConfig
+	var (
+		nodeNetConfig nnc.NodeNetworkConfig
+		ipConfigs     []*cns.ContainerIPConfigState
+		cntxt         context.Context
+		err           error
+	)
 
 	//Get the CRD object
-	if err := r.KubeClient.Get(context.TODO(), request.NamespacedName, &nodeNetConfig); err != nil {
+	cntxt = context.TODO()
+	if err = r.KubeClient.Get(cntxt, request.NamespacedName, &nodeNetConfig); err != nil {
 		if apierrors.IsNotFound(err) {
 			logger.Printf("[cns-rc] CRD not found, ignoring %v", err)
 			return reconcile.Result{}, client.IgnoreNotFound(err)
 		} else {
 			logger.Errorf("[cns-rc] Error retrieving CRD from cache : %v", err)
+			//requeue
 			return reconcile.Result{}, err
 		}
 	}
 
-	logger.Printf("[cns-rc] CRD object: %v", nodeNetConfig)
+	logger.Printf("[cns-rc] CRD spec: %v", nodeNetConfig.Spec)
+	logger.Printf("[cns-rc] CRD status: %v", nodeNetConfig.Status)
 
-	//TODO: Translate CRD status into NetworkContainer request
-	ncRequest, err := CRDStatusToNCRequest(&nodeNetConfig.Status)
+	ipConfigs, err = CRDStatusToCNS(&nodeNetConfig.Status)
 	if err != nil {
 		logger.Errorf("[cns-rc] Error translating crd status to nc request %v", err)
 		//requeue
 		return reconcile.Result{}, err
 	}
 
-	//TODO: process the nc request on CNS side
-	r.CNSClient.UpdateCNSState(ncRequest)
+	if r.CNSClient.ReadyToIPAM() {
+		if err = r.CNSClient.UpdateCNSState(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error updating CNS state: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+	} else {
+		if err = r.updateIPAvailability(cntxt, ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error marking ips as allocated when readying CNS: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+
+		if err = r.CNSClient.InitCNSState(ipConfigs); err != nil {
+			logger.Errorf("[cns-rc] Error initializing cns state: %v", err)
+			//requeue
+			return reconcile.Result{}, err
+		}
+	}
 
 	return reconcile.Result{}, nil
 }
 
 // SetupWithManager Sets up the reconciler with a new manager, filtering using NodeNetworkConfigFilter
 func (r *CrdReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	//Index nodeNames to later be able to filter by nodeName
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{}, "spec.nodeName", func(rawObj runtime.Object) []string {
+		pod := rawObj.(*corev1.Pod)
+		return []string{pod.Spec.NodeName}
+	}); err != nil {
+		return err
+	}
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&nnc.NodeNetworkConfig{}).
-		WithEventFilter(NodeNetworkConfigFilter{nodeName: r.NodeName}).
+		WithEventFilter(NodeNetworkConfigFilter{nodeName: r.NodeName, namespace: r.Namespace}).
```

Comment:
Renme it to ReconcileCNSIpam or ReconcileCNSState, either you should be calling before you initialize Reconcile loop, no? as this is a one time call during initialization. You will list CRD and PODs on a node and reconcileCNS state.